_posts/2019-10-01-intro-to-hmc/hmc-intro.Rmd

```diff
@@ -0,0 +1,283 @@
+
+
+## HMC
+
+So HMC is popular because compared to random-walk-based algorithms, it is a _lot_ more efficient. Unfortunately, it is also a lot more difficult to "get". ^[By "get [it]" I mean the subjective feeling of understanding what's going on; not more and not less.] As discussed in [Math, code, concepts: A third road to deep learning](https://blogs.rstudio.com/tensorflow/posts/2019-03-15-concepts-way-to-dl/), there seem to be (at least) three languages to express an algorithm: Math; code (including pseudo-code, which may or may not be on the verge to math notation); and one I call _conceptual_ which spans the whole range from very abstract to very concrete, even visual. To me personally, HMC is different from most other cases in that even though I find the conceptual explanations fascinating, they result in less "perceived understanding" than either the equations or the code. For people with backgrounds in physics, statistical mechanics and/or differential geometry this will probably be different!
+
+In any case, physical analogies make for the best start.
+
+## Physical analogies
+
+The classic physical analogy is given in the reference article, Radford Neal's "MCMC using Hamiltonian dynamics" [@2012arXiv1206.1901N], and nicely explained in a [video by Ben Lambert](https://www.youtube.com/watch?v=a-wydhEuAm0).
+
+So there's this "thing" we want to maximize, the loglikelihood of the data under the model parameters. Alternatively we can say, we want to minimize the negative loglikelihood (like loss in a neural network). This "thing" to be optimized can then be visualized as an object sliding over a landscape with hills and valleys, and like with gradient descent in deep learning, we want it to end up deep down in some valley. 
+
+In Neal's own words
+
+
+> In two dimensions, we can visualize the dynamics as that of a frictionless puck that slides over a surface of varying height. The state of this system consists of the position of the puck, given by a 2D vector q, and the momentum of the puck (its mass times its velocity), given by a 2D vector p.
+
+Now when you hear "momentum" (and given that I've primed you to think of deep learning) you may feel that sounds familiar, but even though the respective analogies are related the association does not help that much. In deep learning, momentum is commonly praised for its avoidance of ineffective oscillations in imbalanced optimization landscapes ^[Though see [this great distill.pub post](https://distill.pub/2017/momentum/) for a different view.].
+With HMC however, the focus is on the concept of _energy_.
+
+In [statistical mechanics](https://en.wikipedia.org/wiki/Statistical_mechanics), the probability of being in some state is inverse-exponentially related to its energy: ^[Here $T$ is the _temperature_; we won't focus on this so just imagine it being set to 1 in this and subsequent equations.]
```

Comment:
```suggestion
In [statistical mechanics](https://en.wikipedia.org/wiki/Statistical_mechanics), the probability of being in some state  $i$ is inverse-exponentially related to its energy: ^[Here $T$ is the _temperature_; we won't focus on this so just imagine it being set to 1 in this and subsequent equations.]
```

Also consider bringing the footnote out into the main text in parentheses? I missed it the first time and wondered what T was.

Reply:
Makes sense, will do!

Reply:
GH didn't let me commit the suggestion, fixed manually